File: jar-common/src/main/java/com/beancounter/common/model/QuantityValues.java
package com.beancounter.common.model;

import java.math.BigDecimal;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.RequiredArgsConstructor;


/**
 * Accumulation of quantities.
 *
 * @author mikeh
 * @since 2019-01-28
 */
@Data
@Builder
@RequiredArgsConstructor
@AllArgsConstructor
public class QuantityValues {
  @Builder.Default
  private BigDecimal sold = BigDecimal.ZERO;
  @Builder.Default
  private BigDecimal purchased = BigDecimal.ZERO;

  @Builder.Default
  private BigDecimal adjustment = BigDecimal.ZERO;

  private Integer precision;
  private BigDecimal total;

  public BigDecimal getTotal() {
    return (purchased.add(sold)).add(adjustment);
  }

  public Integer getPrecision() {
    // This is a bit hacky. Should be derived from the asset and set not computed
    if (getTotal() == null) {
      return 0;
    }
    if (precision != null) {
      return precision;
    }
    return (getTotal().remainder(BigDecimal.ONE).compareTo(BigDecimal.ZERO) == 0 ? 0 : 3);
  }

}

File: jar-common/src/main/kotlin/com/beancounter/common/model/QuantityValues.kt
package com.beancounter.common.model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import java.math.BigDecimal

/**
 * Accumulation of quantities.
 *
 * @author mikeh
 * @since 2019-01-28
 */
class QuantityValues(
    var sold: BigDecimal = BigDecimal.ZERO,
    var purchased: BigDecimal = BigDecimal.ZERO,
) {
    var adjustment: BigDecimal = BigDecimal.ZERO
    private var precision: Int? = null

    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    fun getTotal(): BigDecimal = purchased.add(sold).add(adjustment)

    @JsonIgnore
    fun hasPosition() = getTotal().compareTo(BigDecimal.ZERO) != 0

    fun getPrecision(): Int {
        // This is a bit hacky. Should be derived from the asset and set not computed
        if (precision != null) {
            return precision as Int
        }
        return if (getTotal().remainder(BigDecimal.ONE).compareTo(BigDecimal.ZERO) == 0) 0 else 3
    }

    fun setPrecision(precision: Int = 2) {
        this.precision = precision
    }

    override fun toString(): String = getTotal().toString()
}
